src/CoffeeNation.Repository/UserLocationRepository.cs

```diff
@@ -0,0 +1,21 @@
+﻿using System.Threading.Tasks;
+using CoffeeNation.Core.Entities;
+using CoffeeNation.Data.Interfaces;
+using CoffeeNation.Repository.Interfaces;
+
+namespace CoffeeNation.Repository
+{
+    public class UserLocationRepository : IUserLocationRepository
+    {
+        private readonly IUserLocationDataReader _dataReader;
+        public UserLocationRepository(IUserLocationDataReader dataReader)
+        {
+            _dataReader = dataReader;
+        }
+
+        public async Task<Location> GetUserLocation()
+        {
+            return await _dataReader.ReadUserLocation();
```

Comment:
Please help me understand the abstractions used and what is the responsability of each abstraction
- Repository
- Data Reader/Writer
- Input / Output Provider

From what i understand, the Repository delegates to the Data Reader/Writer
The data Reader/Writer does the serialization/deserialization of objects from/to a raw I/O (CsvContentProvider / UserLocationProvider). These are concrete per each repository.
The provider uses native apis and is the edge of the system.

Reply:
Everything you said above is exactly so. The providers are the edge of the system, they allow interaction with hard dependencies and also enable us to fake the interactions with the outside of the system. The readers/writers convert the input/output to concepts we can use to operate with, exactly like you said, serialization/deserialization and the repository allows us to define common selection criteria for the deserialized objects, so that the service layer can apply application/business logic (core), without the concern (and without the need to know how to) of querying an entire data set.

Reply:
I've attached a very basic application flow diagram
![coffee_nation_diagram](https://user-images.githubusercontent.com/60556980/74231971-7ea6f180-4cd0-11ea-9f72-5f191683590b.jpg)